<a href="https://colab.research.google.com/github/pranshudatta25/IFN646-portfolio4/blob/main/N11354631_Portfolio4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IFN646 Portfolio Item 4
Submitted By - Pranshu Datta (Student ID - N11354631)

# Installing and Importing Necessary Packages

In [1]:
pip install calour==2019.5.1

In [2]:
!pip install -e git+https://github.com/amnona/dbbact-calour.git#egg=dbbact-calour

Obtaining dbbact-calour from git+https://github.com/amnona/dbbact-calour.git#egg=dbbact-calour
  Updating ./src/dbbact-calour clone
  Running command git fetch -q --tags
  Running command git reset --hard -q 283201acd2d3eba667d7b454ceefba8f3da32a31
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
  Building editable for dbbact-calour (pyproject.toml) ... done
  Created wheel for dbbact-calour: filename=dbbact_calour-1.17.3-0.editable-py3-none-any.whl size=15587 sha256=9f89708f249faf94f018337e5d525f6a34a8c48f16bbe5ca5d1a687c741ace46
  Stored in directory: /tmp/pip-ephem-wheel-cache-ocaw12y0/wheels/88/f0/ca/96aaffd3ceac0aea4d1904eed739b2872b4caf1cc9b788fc93
Successfully built dbbact-calour
  Attempting uninstall: dbbact-calour
    Found existing installation: dbbact-calour 1.17.3
    Uninstalling dbbact-calour-1.17.3:
      Success

In [3]:
!pip install biom-format

In [4]:
import calour as ca
ca.set_log_level(11)

import biom
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

# Data Loading & Exploration

In [5]:
biom_table = biom.load_table('/content/otu_table.biom')

In [6]:
feature_ids = biom_table.ids(axis='observation')
print(feature_ids)

['547831' '206494' '4333043' ... '279470' '279473' '4436135']


In [7]:
sample_ids = biom_table.ids(axis='sample')
print(sample_ids[:10])

['524.ESC.1.65.NPR' '524.ESC.1.23a.OPR' '524.ESC.1.33.OPR'
 '524.ESC.1.53.OPL' '524.ESC.1.72.OPL' '524.ESC.1.1.OPR'
 '524.ESC.1.11.NPL' '524.ESC.1.12.OPR' '524.ESC.1.13.OPL'
 '524.ESC.1.14.OPL']


In [8]:
print(biom_table)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [52]:
warnings.filterwarnings("ignore", message="These have metadata but do not have data", category=UserWarning)

# Your code to read the amplicon data
data = ca.read_amplicon('/content/otu_table.biom',
                      sample_metadata_file='/content/metadata.txt',
                      feature_metadata_file='/content/97_otu_taxonomy.txt',
                      normalize=10000,
                      min_reads=1000)

2023-10-17 08:48:59 INFO loaded 290 samples, 4572 features
2023-10-17 08:48:59 WARNING These have metadata but do not have data - dropped (1): {'524.ESC.1.54.OPR'}
2023-10-17 08:49:00 WARNING These have metadata but do not have data - dropped (94749): {'3475526', '4345360', '588988', '548924', '573498', '242024', '4399805', '2453', '202207', '4433022', '1147576', '203694', '107161', '136218', '40838', '297116', '586689', '329367', '4303318', '101033', '1961379', '353214', '342277', '4480117', '1106054', '3461862', '772190', '4412860', '16115', '352943', '796173', '2457426', '255856', '2811009', '4341233', '195470', '1106325', '579148', '1141793', '240045', '586019', '1037111', '968078', '328146', '822693', '4364008', '157481', '1631024', '570132', '851821', '162651', '156126', '532759', '4357222', '185802', '278172', '2904385', '540671', '512178', '56514', '2170536', '4381149', '544746', '330757', '2592837', '1049851', '4346003', '4470519', '543661', '824904', '4333759', '3371208', '58

The data file of BIOM type are loaded in conjunction with suitable sample & feature metadata files. Data is also normalized and samples with less than 1000 reads are removed.

In [10]:
print(data)

AmpliconExperiment with 272 samples, 4572 features


The result is an AmpliconExperiment with 272 samples and 4572 features.

In [11]:
data=data.filter_sum_abundance(10)

2023-10-17 08:44:55 INFO After filtering, 2171 remain.


Non-interesting bacteria, i.e., bacteria with less than 10 (normalized) reads total over all samples have been removed from this analysis.



In [12]:
data

AmpliconExperiment with 272 samples, 2171 features

The result is an AmpliconExperiment with 272 samples and lesser number of features, i.e., 2171 features.

In [13]:
df = pd.read_csv('/content/metadata.txt', sep='\t').head(20)
df.head(5).T

0  \
#SampleID                             524.ESC.1.22.NPR   
age                                                 24   
airwaysite                                        Nose   
antibioticusepast3months         AntibioticUsed.1month   
chronic_condition                                  Yes   
host_subject_id                               ESC.1.22   
respiratorydiseasestatus  NasalCongestion.mild.current   
sex                                               male   
sideofbody                                       Right   
smoker                                             Yes   
Description                                       Nose   

                                                     1  \
#SampleID                             524.ESC.1.22.OPL   
age                                                 24   
airwaysite                                      Throat   
antibioticusepast3months         AntibioticUsed.1month   
chronic_condition                                  Yes   
host_subject_id                               ESC.1.22   
respiratorydiseasestatus  NasalCongestion.mild.current   
sex                                               male   
sideofbody                                        Left   
smoker                                             Yes   
Description                                     Throat   

                                                     2  \
#SampleID                             524.ESC.1.22.OPR   
age                                                 24   
airwaysite                                      Throat   
antibioticusepast3months         AntibioticUsed.1month   
chronic_condition                                  Yes   
host_subject_id                               ESC.1.22   
respiratorydiseasestatus  NasalCongestion.mild.current   
sex                                               male   
sideofbody                                       Right   
smoker                                             Yes   
Description                                     Throat   

                                                         3  \
#SampleID                                 524.ESC.1.23.NPL   
age                                                     24   
airwaysite                                            Nose   
antibioticusepast3months                              None   
chronic_condition                                      Yes   
host_subject_id                                   ESC.1.23   
respiratorydiseasestatus  NasalCongestion.moderate.current   
sex                                                   male   
sideofbody                                            Left   
smoker                                                 Yes   
Description                                           Nose   

                                                         4  
#SampleID                                 524.ESC.1.23.NPR  
age                                                     24  
airwaysite                                            Nose  
antibioticusepast3months                              None  
chronic_condition                                      Yes  
host_subject_id                                   ESC.1.23  
respiratorydiseasestatus  NasalCongestion.moderate.current  
sex                                                   male  
sideofbody                                           Right  
smoker                                                 Yes  
Description                                           Nose

Exploring the provided sample metadata file and analyzing the attributes and their corresponding values.

# Differential Microbial Abundance Analysis

# Task 1

In [14]:
data = data.cluster_features()

2023-10-17 08:44:55 INFO After filtering, 2171 remain.


In [15]:
result = data.diff_abundance('airwaysite','Nose','Throat', random_seed=2018)

2023-10-17 08:44:57 INFO 272 samples with both values
2023-10-17 08:44:57 INFO After filtering, 2171 remain.
2023-10-17 08:44:57 INFO 136 samples with value 1 (['Nose'])
2023-10-17 08:44:58 INFO number of higher in Nose: 594. number of higher in Throat : 601. total 1195


In [16]:
result2 = data.diff_abundance('airwaysite','Nose','Throat', random_seed=2018, fdr_method='bhfdr')

2023-10-17 08:44:58 INFO 272 samples with both values
2023-10-17 08:44:58 INFO After filtering, 2171 remain.
2023-10-17 08:44:58 INFO 136 samples with value 1 (['Nose'])
2023-10-17 08:44:59 INFO number of higher in Nose: 509. number of higher in Throat : 546. total 1055


'result2' will be used for the purpose of further analysis with fdr_method = 'bhfdr'(Benjamini-Hochberg FDR) as it is beneficial in striking a balance between statistical stringency and sensitivity. By controlling the False Discovery Rate, it limits the rate of false positives while maximizing the detection of true positives, mitigating the risk of erroneous results.

In [17]:
result2.feature_metadata

k__Bacteria; p__Bacteroidetes; c__Flavobacteriia; o__Flavobacteriales; f__Flavobacteriaceae; g__Flavobacterium; s__  \
252119   k__Bacteria; p__Actinobacteria; c__Actinobacte...                                                                    
4465561  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
261241   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
342427   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
301574   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
...                                                    ...                                                                    
254021   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
206938   k__Bacteria; p__Proteobacteria; c__Gammaproteo...                                                                    
842413   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
288680   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
200189   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    

        _feature_id                                           taxonomy  \
252119       252119  k__Bacteria;p__Actinobacteria;c__Actinobacteri...   
4465561     4465561  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
261241       261241  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
342427       342427  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
301574       301574  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
...             ...                                                ...   
254021       254021  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
206938       206938  k__Bacteria;p__Proteobacteria;c__Gammaproteoba...   
842413       842413  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
288680       288680  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
200189       200189  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   

         _calour_stat  _calour_pval  _calour_qval _calour_direction  
252119    -122.492647      0.000999      0.003597            Throat  
4465561   -121.544118      0.000999      0.003597            Throat  
261241    -120.595588      0.000999      0.003597            Throat  
342427    -120.066176      0.000999      0.003597            Throat  
301574    -120.036765      0.000999      0.003597            Throat  
...               ...           ...           ...               ...  
254021     115.455882      0.000999      0.003597              Nose  
206938     116.058824      0.000999      0.003597              Nose  
842413     116.352941      0.000999      0.003597              Nose  
288680     118.566176      0.000999      0.003597              Nose  
200189     118.669118      0.000999      0.003597              Nose  

[1055 rows x 7 columns]

In [18]:
result2.feature_metadata['taxonomy']

252119     k__Bacteria;p__Actinobacteria;c__Actinobacteri...
4465561    k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
261241     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
342427     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
301574     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
                                 ...                        
254021     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
206938     k__Bacteria;p__Proteobacteria;c__Gammaproteoba...
842413     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
288680     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
200189     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
Name: taxonomy, Length: 1055, dtype: object

In [19]:
def fetch_taxonomy(taxonomy_str):
    hierarchies = taxonomy_str.split(';')
    hierarchies_dict = {}
    for hierarchy in hierarchies:
        parts = hierarchy.strip().split('__')
        if len(parts) == 2:
            hierarchies_dict[parts[0]] = parts[1]
    return hierarchies_dict

In [20]:
taxonomy_data = result2.feature_metadata[['taxonomy']]
taxonomy_data['taxonomy_string'] = taxonomy_data['taxonomy'].apply(fetch_taxonomy)

taxonomy_data['ID'] = taxonomy_data.index
taxonomy_data['Kingdom'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('k', ''))
taxonomy_data['Phylum'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('p', ''))
taxonomy_data['Class'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('c', ''))
taxonomy_data['Order'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('o', ''))
taxonomy_data['Family'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('f', ''))
taxonomy_data['Genus'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('g', ''))
taxonomy_data['Species'] = taxonomy_data['taxonomy_string'].apply(lambda x: x.get('s', ''))

Calling the fetch_taxonomy() method to retrieve line-wise bacteria taxonomy information using the provided prefix in the guidelines.

In [21]:
def get_taxonomy_name(row):
    genus = row['Genus']
    species = row['Species']
    family = row['Family']
    order = row['Order']
    class_ = row['Class']
    phylum = row['Phylum']

    if genus and species:
        return f"{genus.split('__')[-1]} {species.split('__')[-1]}"
    elif genus:
        return genus.split('__')[-1]
    elif family:
        return family.split('__')[-1]
    elif order:
        return order.split('__')[-1]
    elif class_:
        return class_.split('__')[-1]
    else:
        return phylum.split('__')[-1]

 If both genus and species are available, both are reported. If only genus is available, only it is reported. If neither genus nor species is available, the most informative level available is checked and reported accordingly.

In [22]:
taxonomy_data['Bacteria Name'] = taxonomy_data.apply(get_taxonomy_name, axis=1)

taxonomy_data.drop(columns=['taxonomy_string'], inplace=True)
taxonomy_data.drop(columns=['Genus', 'Species', 'taxonomy', 'Kingdom', 'Phylum', 'ID', 'Class', 'Order', 'Family'], inplace=True)

taxonomy_data = taxonomy_data.rename_axis('OTU')

print(taxonomy_data)

                     Bacteria Name
OTU                               
252119              Micrococcaceae
4465561  Prevotella melaninogenica
261241              Granulicatella
342427          Veillonella dispar
301574               Streptococcus
...                            ...
254021                 Leuconostoc
206938               Acinetobacter
842413                 Leuconostoc
288680                 Lactococcus
200189                 Lactococcus

[1055 rows x 1 columns]


In [23]:
print(f"The bacteria that have a differential abundance between the nose and the throat (irrespective of the smoking status) are -\n")
print("OTU | Bacteria Name \n")
for index, row in taxonomy_data.iterrows():
    print(f"{index} | {row['Bacteria Name']}")

The bacteria that have a differential abundance between the nose and the throat (irrespective of the smoking status) are -

OTU | Bacteria Name 

252119 | Micrococcaceae
4465561 | Prevotella melaninogenica
261241 | Granulicatella
342427 | Veillonella dispar
301574 | Streptococcus
1105219 | Campylobacter
3866487 | Oribacterium
516611 | Streptococcus
4452538 | Fusobacterium
778484 | Actinomyces
977292 | [Prevotella]
652731 | Paludibacter
1887417 | TM7-3
4396516 | Lachnospiraceae
272150 | Streptococcus
4405869 | Fusobacterium
4476186 | Prevotella stercorea
1144463 | Veillonella
274257 | Bulleidia
1019594 | Lachnospiraceae
4393019 | Catonella
218785 | Neisseria
4295238 | [Prevotella]
341460 | Haemophilus
4475758 | Veillonella dispar
518743 | Veillonella parvula
4401373 | Veillonella dispar
4297119 | Leptotrichia
875735 | Actinomyces
258034 | Selenomonas
4451251 | Atopobium
523036 | Gemellaceae
4447605 | Lachnospiraceae
1903527 | Megasphaera
2171537 | Porphyromonas
349539 | F16
259015 | Vei

# Task 2

In [24]:
result3 = data.diff_abundance('smoker','Yes','No', random_seed=2018, fdr_method='bhfdr')

2023-10-17 08:45:00 INFO 272 samples with both values
2023-10-17 08:45:00 INFO After filtering, 2171 remain.
2023-10-17 08:45:00 INFO 135 samples with value 1 (['Yes'])
2023-10-17 08:45:03 INFO number of higher in Yes: 54. number of higher in No : 15. total 69


Just as Task 1, fdr_method = 'bhfdr' has been opted as it effectively mitigates risk of erroneous or false positive results.

In [25]:
result3.feature_metadata

k__Bacteria; p__Bacteroidetes; c__Flavobacteriia; o__Flavobacteriales; f__Flavobacteriaceae; g__Flavobacterium; s__  \
307401   k__Bacteria; p__Bacteroidetes; c__Cytophagia; ...                                                                    
218785   k__Bacteria; p__Proteobacteria; c__Betaproteob...                                                                    
269755   k__Bacteria; p__Actinobacteria; c__Actinobacte...                                                                    
557665   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
4361797  k__Bacteria; p__Bacteroidetes; c__Flavobacteri...                                                                    
...                                                    ...                                                                    
178559   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
178631   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
179473   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
178629   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
178627   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    

        _feature_id                                           taxonomy  \
307401       307401  k__Bacteria;p__Bacteroidetes;c__Cytophagia;o__...   
218785       218785  k__Bacteria;p__Proteobacteria;c__Betaproteobac...   
269755       269755  k__Bacteria;p__Actinobacteria;c__Actinobacteri...   
557665       557665  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
4361797     4361797  k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...   
...             ...                                                ...   
178559       178559  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
178631       178631  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
179473       179473  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
178629       178629  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
178627       178627  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   

         _calour_stat  _calour_pval  _calour_qval _calour_direction  
307401     -31.038443      0.000999      0.051639                No  
218785     -29.582482      0.001998      0.077458                No  
269755     -28.332414      0.000999      0.051639                No  
557665     -27.244120      0.000999      0.051639                No  
4361797    -26.597026      0.002997      0.094297                No  
...               ...           ...           ...               ...  
178559      31.612003      0.000999      0.051639               Yes  
178631      32.817951      0.000999      0.051639               Yes  
179473      33.200324      0.000999      0.051639               Yes  
178629      34.700406      0.000999      0.051639               Yes  
178627      43.509705      0.000999      0.051639               Yes  

[69 rows x 7 columns]

In [26]:
result3.feature_metadata['taxonomy']

307401     k__Bacteria;p__Bacteroidetes;c__Cytophagia;o__...
218785     k__Bacteria;p__Proteobacteria;c__Betaproteobac...
269755     k__Bacteria;p__Actinobacteria;c__Actinobacteri...
557665     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
4361797    k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...
                                 ...                        
178559     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
178631     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
179473     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
178629     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
178627     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
Name: taxonomy, Length: 69, dtype: object

In [27]:
taxonomy_data_task2 = result3.feature_metadata[['taxonomy']]
taxonomy_data_task2['taxonomy_string'] = taxonomy_data_task2['taxonomy'].apply(fetch_taxonomy)

taxonomy_data_task2['ID'] = taxonomy_data_task2.index
taxonomy_data_task2['Kingdom'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('k', ''))
taxonomy_data_task2['Phylum'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('p', ''))
taxonomy_data_task2['Class'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('c', ''))
taxonomy_data_task2['Order'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('o', ''))
taxonomy_data_task2['Family'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('f', ''))
taxonomy_data_task2['Genus'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('g', ''))
taxonomy_data_task2['Species'] = taxonomy_data_task2['taxonomy_string'].apply(lambda x: x.get('s', ''))

In [28]:
taxonomy_data_task2['Bacteria Name'] = taxonomy_data_task2.apply(get_taxonomy_name, axis=1)

taxonomy_data_task2.drop(columns=['taxonomy_string'], inplace=True)
taxonomy_data_task2.drop(columns=['Genus', 'Species', 'taxonomy', 'Kingdom', 'Phylum', 'ID', 'Class', 'Order', 'Family'], inplace=True)

taxonomy_data_task2 = taxonomy_data_task2.rename_axis('OTU')

print(taxonomy_data_task2)

                Bacteria Name
OTU                          
307401      Cyclobacteriaceae
218785              Neisseria
269755   Candidatus Rhodoluna
557665             Prevotella
4361797        Flavobacterium
...                       ...
178559            Coprococcus
178631          Clostridiales
179473                Blautia
178629            Coprococcus
178627        Lachnospiraceae

[69 rows x 1 columns]


In [29]:
print(f"The bacteria that have a differential abundance between smokers and non-smokers (irrespective of the sampling site) are -\n")
print("OTU | Bacteria Name \n")
for index, row in taxonomy_data_task2.iterrows():
    print(f"{index} | {row['Bacteria Name']}")

The bacteria that have a differential abundance between smokers and non-smokers (irrespective of the sampling site) are -

OTU | Bacteria Name 

307401 | Cyclobacteriaceae
218785 | Neisseria
269755 | Candidatus Rhodoluna
557665 | Prevotella
4361797 | Flavobacterium
4443201 | Tannerella
4454893 | Flavobacterium
3297209 | Rhodocyclaceae
171202 | Prevotella
332355 | Zoogloea
301251 | Prevotella
1036557 | Neisseria subflava
677680 | [Prevotella]
223957 | Ralstonia
557093 | Enterobacteriaceae
95168 | Treponema
182122 | Lachnospiraceae
583522 | Veillonellaceae
128382 | Dialister
188974 | Coprococcus
181834 | Lachnospiraceae
255809 | Alicyclobacillaceae
178859 | Ruminococcus
253550 | Bacteroidales
583500 | Dorea
551179 | Coriobacteriaceae
190633 | Blautia
4463709 | Coprococcus
192619 | Lachnospiraceae
180155 | Ruminococcaceae
191754 | Oscillospira
3931 | Treponema
294791 | Coprococcus
849723 | Erysipelotrichaceae
4481084 | Clostridiaceae
178621 | Coprococcus
834424 | Treponema
299439 | Prevot

# Task 3

# Part 1

In [30]:
data_nose = data.filter_samples('airwaysite', 'Nose')

filter_samples() method is used to filter out records where airwaysite = 'Nose'.

In [31]:
result4 = data_nose.diff_abundance('smoker','Yes','No', random_seed=2018, fdr_method='bhfdr')

2023-10-17 08:45:04 INFO 136 samples with both values
2023-10-17 08:45:04 INFO After filtering, 1792 remain.
2023-10-17 08:45:04 INFO 67 samples with value 1 (['Yes'])
2023-10-17 08:45:06 INFO number of higher in Yes: 61. number of higher in No : 13. total 74


To maintain consistency in the experiment, fdr_method = 'bhfdr' is opted for this and the remaining tasks.

In [32]:
result4.feature_metadata

k__Bacteria; p__Bacteroidetes; c__Flavobacteriia; o__Flavobacteriales; f__Flavobacteriaceae; g__Flavobacterium; s__  \
254021   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
4227110  k__Bacteria; p__Proteobacteria; c__Gammaproteo...                                                                    
252679   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
146907   k__Bacteria; p__Proteobacteria; c__Gammaproteo...                                                                    
3944484  k__Bacteria; p__Proteobacteria; c__Gammaproteo...                                                                    
...                                                    ...                                                                    
178631   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
179473   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
516611   k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...                                                                    
178629   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    
178627   k__Bacteria; p__Firmicutes; c__Clostridia; o__...                                                                    

        _feature_id                                           taxonomy  \
254021       254021  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
4227110     4227110  k__Bacteria;p__Proteobacteria;c__Gammaproteoba...   
252679       252679  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
146907       146907  k__Bacteria;p__Proteobacteria;c__Gammaproteoba...   
3944484     3944484  k__Bacteria;p__Proteobacteria;c__Gammaproteoba...   
...             ...                                                ...   
178631       178631  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
179473       179473  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
516611       516611  k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
178629       178629  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   
178627       178627  k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...   

         _calour_stat  _calour_pval  _calour_qval _calour_direction  
254021     -25.255462      0.000999      0.035102                No  
4227110    -25.196626      0.000999      0.035102                No  
252679     -25.049535      0.000999      0.035102                No  
146907     -24.534718      0.000999      0.035102                No  
3944484    -23.843392      0.001998      0.058695                No  
...               ...           ...           ...               ...  
178631      27.108804      0.000999      0.035102               Yes  
179473      27.873675      0.000999      0.035102               Yes  
516611      28.079602      0.000999      0.035102               Yes  
178629      28.079602      0.000999      0.035102               Yes  
178627      31.139087      0.000999      0.035102               Yes  

[74 rows x 7 columns]

In [33]:
result4.feature_metadata['taxonomy']

254021     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
4227110    k__Bacteria;p__Proteobacteria;c__Gammaproteoba...
252679     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
146907     k__Bacteria;p__Proteobacteria;c__Gammaproteoba...
3944484    k__Bacteria;p__Proteobacteria;c__Gammaproteoba...
                                 ...                        
178631     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
179473     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
516611     k__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...
178629     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
178627     k__Bacteria;p__Firmicutes;c__Clostridia;o__Clo...
Name: taxonomy, Length: 74, dtype: object

In [34]:
taxonomy_data_task3_pt1 = result4.feature_metadata[['taxonomy']]
taxonomy_data_task3_pt1['taxonomy_string'] = taxonomy_data_task3_pt1['taxonomy'].apply(fetch_taxonomy)

taxonomy_data_task3_pt1['ID'] = taxonomy_data_task3_pt1.index
taxonomy_data_task3_pt1['Kingdom'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('k', ''))
taxonomy_data_task3_pt1['Phylum'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('p', ''))
taxonomy_data_task3_pt1['Class'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('c', ''))
taxonomy_data_task3_pt1['Order'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('o', ''))
taxonomy_data_task3_pt1['Family'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('f', ''))
taxonomy_data_task3_pt1['Genus'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('g', ''))
taxonomy_data_task3_pt1['Species'] = taxonomy_data_task3_pt1['taxonomy_string'].apply(lambda x: x.get('s', ''))

In [35]:
taxonomy_data_task3_pt1['Bacteria Name'] = taxonomy_data_task3_pt1.apply(get_taxonomy_name, axis=1)

taxonomy_data_task3_pt1.drop(columns=['taxonomy_string'], inplace=True)
taxonomy_data_task3_pt1.drop(columns=['Genus', 'Species', 'taxonomy', 'Kingdom', 'Phylum', 'ID', 'Class', 'Order', 'Family'], inplace=True)

taxonomy_data_task3_pt1 = taxonomy_data_task3_pt1.rename_axis('OTU')

print(taxonomy_data_task3_pt1)

                Bacteria Name
OTU                          
254021            Leuconostoc
4227110  Citrobacter freundii
252679     Leuconostoc fallax
146907          Enhydrobacter
3944484    Enterobacteriaceae
...                       ...
178631          Clostridiales
179473                Blautia
516611          Streptococcus
178629            Coprococcus
178627        Lachnospiraceae

[74 rows x 1 columns]


In [36]:
print(f"The bacteria that have a differential abundance in the nose between smokers and non-smokers are -\n")
print("OTU | Bacteria Name \n")
for index, row in taxonomy_data_task3_pt1.iterrows():
    print(f"{index} | {row['Bacteria Name']}")

The bacteria that have a differential abundance in the nose between smokers and non-smokers are -

OTU | Bacteria Name 

254021 | Leuconostoc
4227110 | Citrobacter freundii
252679 | Leuconostoc fallax
146907 | Enhydrobacter
3944484 | Enterobacteriaceae
288680 | Lactococcus
206938 | Acinetobacter
539924 | Zoogloea
4307097 | [Weeksellaceae]
4361797 | Flavobacterium
332355 | Zoogloea
4454893 | Flavobacterium
557093 | Enterobacteriaceae
1132297 | Lactobacillus zeae
237945 | Oribacterium
1076668 | Aerococcaceae
205118 | RF16
250269 | Veillonellaceae
1042479 | Prevotella melaninogenica
182122 | Lachnospiraceae
46931 | Brenneria salicis
128382 | Dialister
183824 | Blautia
181834 | Lachnospiraceae
188974 | Coprococcus
192619 | Lachnospiraceae
4463709 | Coprococcus
178859 | Ruminococcus
255809 | Alicyclobacillaceae
190633 | Blautia
178621 | Coprococcus
583500 | Dorea
294791 | Coprococcus
182497 | Clostridiales
849723 | Erysipelotrichaceae
191754 | Oscillospira
184882 | Blautia
4479397 | Bactero

# Part 2

In [37]:
data_throat = data.filter_samples('airwaysite', 'Throat')

filter_samples() method is used to filter out records where airwaysite = 'Throat'.

In [38]:
result5 = data_throat.diff_abundance('smoker','Yes','No', random_seed=2018, fdr_method='bhfdr')

2023-10-17 08:45:06 INFO 136 samples with both values
2023-10-17 08:45:06 INFO After filtering, 1682 remain.
2023-10-17 08:45:06 INFO 68 samples with value 1 (['Yes'])
2023-10-17 08:45:08 INFO number of higher in Yes: 2. number of higher in No : 15. total 17


In [39]:
result5.feature_metadata

k__Bacteria; p__Bacteroidetes; c__Flavobacteriia; o__Flavobacteriales; f__Flavobacteriaceae; g__Flavobacterium; s__  \
218785   k__Bacteria; p__Proteobacteria; c__Betaproteob...                                                                    
4433523  k__Bacteria; p__Bacteroidetes; c__Flavobacteri...                                                                    
347882   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
557665   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
341460   k__Bacteria; p__Proteobacteria; c__Gammaproteo...                                                                    
4452538  k__Bacteria; p__Fusobacteria; c__Fusobacteriia...                                                                    
1042479  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
3297209  k__Bacteria; p__Proteobacteria; c__Betaproteob...                                                                    
4484389  k__Bacteria; p__Proteobacteria; c__Betaproteob...                                                                    
4451646  k__Bacteria; p__Bacteroidetes; c__Flavobacteri...                                                                    
4374639  k__Bacteria; p__Proteobacteria; c__Gammaproteo...                                                                    
3291558  k__Bacteria; p__Proteobacteria; c__Epsilonprot...                                                                    
301251   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
171202   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
677680   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
253550   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...                                                                    
4451251  k__Bacteria; p__Actinobacteria; c__Coriobacter...                                                                    

        _feature_id                                           taxonomy  \
218785       218785  k__Bacteria;p__Proteobacteria;c__Betaproteobac...   
4433523     4433523  k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...   
347882       347882  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
557665       557665  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
341460       341460  k__Bacteria;p__Proteobacteria;c__Gammaproteoba...   
4452538     4452538  k__Bacteria;p__Fusobacteria;c__Fusobacteriia;o...   
1042479     1042479  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
3297209     3297209  k__Bacteria;p__Proteobacteria;c__Betaproteobac...   
4484389     4484389  k__Bacteria;p__Proteobacteria;c__Betaproteobac...   
4451646     4451646  k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...   
4374639     4374639  k__Bacteria;p__Proteobacteria;c__Gammaproteoba...   
3291558     3291558  k__Bacteria;p__Proteobacteria;c__Epsilonproteo...   
301251       301251  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
171202       171202  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
677680       677680  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
253550       253550  k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...   
4451251     4451251  k__Bacteria;p__Actinobacteria;c__Coriobacterii...   

         _calour_stat  _calour_pval  _calour_qval _calour_direction  
218785     -34.117647      0.000999      0.098842                No  
4433523    -31.000000      0.000999      0.098842                No  
347882     -30.852941      0.000999      0.098842                No  
557665     -27.911765      0.000999      0.098842                No  
341460     -26.294118      0.000999   

In [40]:
result5.feature_metadata['taxonomy']

218785     k__Bacteria;p__Proteobacteria;c__Betaproteobac...
4433523    k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...
347882     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
557665     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
341460     k__Bacteria;p__Proteobacteria;c__Gammaproteoba...
4452538    k__Bacteria;p__Fusobacteria;c__Fusobacteriia;o...
1042479    k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
3297209    k__Bacteria;p__Proteobacteria;c__Betaproteobac...
4484389    k__Bacteria;p__Proteobacteria;c__Betaproteobac...
4451646    k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...
4374639    k__Bacteria;p__Proteobacteria;c__Gammaproteoba...
3291558    k__Bacteria;p__Proteobacteria;c__Epsilonproteo...
301251     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
171202     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
677680     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
253550     k__Bacteria;p__Bacteroidetes;c__Bacteroidia;o_...
4451251    k__Bacteria;p

In [41]:
taxonomy_data_task3_pt2 = result5.feature_metadata[['taxonomy']]
taxonomy_data_task3_pt2['taxonomy_string'] = taxonomy_data_task3_pt2['taxonomy'].apply(fetch_taxonomy)

taxonomy_data_task3_pt2['ID'] = taxonomy_data_task3_pt2.index
taxonomy_data_task3_pt2['Kingdom'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('k', ''))
taxonomy_data_task3_pt2['Phylum'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('p', ''))
taxonomy_data_task3_pt2['Class'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('c', ''))
taxonomy_data_task3_pt2['Order'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('o', ''))
taxonomy_data_task3_pt2['Family'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('f', ''))
taxonomy_data_task3_pt2['Genus'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('g', ''))
taxonomy_data_task3_pt2['Species'] = taxonomy_data_task3_pt2['taxonomy_string'].apply(lambda x: x.get('s', ''))

In [42]:
taxonomy_data_task3_pt2['Bacteria Name'] = taxonomy_data_task3_pt2.apply(get_taxonomy_name, axis=1)

taxonomy_data_task3_pt2.drop(columns=['taxonomy_string'], inplace=True)
taxonomy_data_task3_pt2.drop(columns=['Genus', 'Species', 'taxonomy', 'Kingdom', 'Phylum', 'ID', 'Class', 'Order', 'Family'], inplace=True)

taxonomy_data_task3_pt2 = taxonomy_data_task3_pt2.rename_axis('OTU')

print(taxonomy_data_task3_pt2)

                      Bacteria Name
OTU                                
218785                    Neisseria
4433523              Capnocytophaga
347882                         RF16
557665                   Prevotella
341460                  Haemophilus
4452538               Fusobacterium
1042479   Prevotella melaninogenica
3297209              Rhodocyclaceae
4484389   Bergeriella denitrificans
4451646              Capnocytophaga
4374639  Haemophilus parainfluenzae
3291558               Campylobacter
301251                   Prevotella
171202                   Prevotella
677680                 [Prevotella]
253550                Bacteroidales
4451251                   Atopobium


In [43]:
print(f"The bacteria that have a differential abundance in the throat between smokers and non-smokers are -\n")
print("OTU | Bacteria Name \n")
for index, row in taxonomy_data_task3_pt2.iterrows():
    print(f"{index} | {row['Bacteria Name']}")

The bacteria that have a differential abundance in the throat between smokers and non-smokers are -

OTU | Bacteria Name 

218785 | Neisseria
4433523 | Capnocytophaga
347882 | RF16
557665 | Prevotella
341460 | Haemophilus
4452538 | Fusobacterium
1042479 | Prevotella melaninogenica
3297209 | Rhodocyclaceae
4484389 | Bergeriella denitrificans
4451646 | Capnocytophaga
4374639 | Haemophilus parainfluenzae
3291558 | Campylobacter
301251 | Prevotella
171202 | Prevotella
677680 | [Prevotella]
253550 | Bacteroidales
4451251 | Atopobium


# Part 3

In [44]:
taxonomy_data_task3_combined = pd.concat([taxonomy_data_task3_pt1, taxonomy_data_task3_pt2], axis=0, join='outer')
taxonomy_data_task3_combined

Bacteria Name
OTU                          
254021            Leuconostoc
4227110  Citrobacter freundii
252679     Leuconostoc fallax
146907          Enhydrobacter
3944484    Enterobacteriaceae
...                       ...
301251             Prevotella
171202             Prevotella
677680           [Prevotella]
253550          Bacteroidales
4451251             Atopobium

[91 rows x 1 columns]

In [45]:
duplicated = taxonomy_data_task3_combined[taxonomy_data_task3_combined.index.duplicated(keep=False)]
duplicated

Bacteria Name
OTU                               
1042479  Prevotella melaninogenica
341460                 Haemophilus
341460                 Haemophilus
1042479  Prevotella melaninogenica

In [46]:
taxonomy_data_task3_combined = taxonomy_data_task3_combined[~taxonomy_data_task3_combined.index.duplicated(keep=False)]
taxonomy_data_task3_combined

Bacteria Name
OTU                          
254021            Leuconostoc
4227110  Citrobacter freundii
252679     Leuconostoc fallax
146907          Enhydrobacter
3944484    Enterobacteriaceae
...                       ...
301251             Prevotella
171202             Prevotella
677680           [Prevotella]
253550          Bacteroidales
4451251             Atopobium

[87 rows x 1 columns]

In the above lines of code, dataframes with bacteria detail having differential abundance between smokers and non-smokers in nose (Task 3, Part 1) and in throat (Task 3, Part 2) are combined. Any repeated OTU ID has been removed accordingly.

In [47]:
taxonomy_data_task2.shape

(69, 1)

The dataframe 'taxonomy_data_task2' which is a list of OTU IDs and bacteria names where there is differential abundance between smokers and non-smokers (irrespective of the sampling site). There are a total of 69 records in the dataframe 'taxonomy_data_task2'.

In [48]:
taxonomy_data_task3_combined.shape

(87, 1)

Whereas, 'taxonomy_data_task3_combined' is a list that has been created by initially identifying bacteria with differential abundance in the nose between smokers and non-smokers and separately identifying bacteria with differential abundance in the throat between smokers and non-smokers. There are a total of 87 records in the dataframe 'taxonomy_data_task3_combined'.

Judging from the differing number of records in both the dataframes, it can be said that the results obtained in Task 3 are not the same as the results in Task 2.

In [49]:
common_records = taxonomy_data_task2.join(
    taxonomy_data_task3_combined, how='inner', lsuffix='_task2', rsuffix='_task3_combined')
common_records

Bacteria Name_task2  Bacteria Name_task3_combined
OTU                                                                
218785                      Neisseria                     Neisseria
557665                     Prevotella                    Prevotella
4361797                Flavobacterium                Flavobacterium
4454893                Flavobacterium                Flavobacterium
3297209                Rhodocyclaceae                Rhodocyclaceae
171202                     Prevotella                    Prevotella
332355                       Zoogloea                      Zoogloea
301251                     Prevotella                    Prevotella
677680                   [Prevotella]                  [Prevotella]
557093             Enterobacteriaceae            Enterobacteriaceae
182122                Lachnospiraceae               Lachnospiraceae
128382                      Dialister                     Dialister
188974                    Coprococcus                   Coprococcus
181834                Lachnospiraceae               Lachnospiraceae
255809            Alicyclobacillaceae           Alicyclobacillaceae
178859                   Ruminococcus                  Ruminococcus
253550                  Bacteroidales                 Bacteroidales
583500                          Dorea                         Dorea
190633                        Blautia                       Blautia
4463709                   Coprococcus                   Coprococcus
192619                Lachnospiraceae               Lachnospiraceae
191754                   Oscillospira                  Oscillospira
294791                    Coprococcus                   Coprococcus
849723            Erysipelotrichaceae           Erysipelotrichaceae
4481084                Clostridiaceae                Clostridiaceae
178621                    Coprococcus                   Coprococcus
992686                      WAL_1855D                     WAL_1855D
184882                        Blautia                       Blautia
185190                Lachnospiraceae               Lachnospiraceae
753638                      Dialister                     Dialister
178882                    Coprococcus                   Coprococcus
182497                  Clostridiales                 Clostridiales
180687                Ruminococcaceae               Ruminococcaceae
179319                          Dorea                         Dorea
179796                Lachnospiraceae               Lachnospiraceae
178926                  Clostridiales                 Clostridiales
178665                Lachnospiraceae               Lachnospiraceae
185667                    Coprococcus                   Coprococcus
4373156               Lachnospiraceae               Lachnospiraceae
257492                  Aerococcaceae                 Aerococcaceae
179665          Dorea formicigenerans         Dorea formicigenerans
751601             Lactobacillus zeae            Lactobacillus zeae
179408                 [Ruminococcus]                [Ruminococcus]
4393532             Eggerthella lenta             Eggerthella lenta
178843                  Clostridiales                 Clostridiales
2576209  Pseudoramibacter_Eubacterium  Pseudoramibacter_Eubacterium
4484034                   Bacteroides                   Bacteroides
196572                  Clostridiales                 Clostridiales
4396877        [Eubacterium] dolichum        [Eubacterium] dolichum
4480970          Bacteroides fragilis          Bacteroides fragilis
197105          [Eubacterium] biforme         [Eubacterium] biforme
178559                    Coprococcus                   Coprococcus
178631                  Clostridiales                 Clostridiales
179473                        Blautia                       Blautia
178629                    Coprococcus                   Coprococcus
178627                Lachnospiraceae               Lachnospiraceae

In [50]:
common_records.shape

(56, 2)

Based on the results of the inner join, it can be concluded that there are a total of 56 common bacterial names that exhibit differential abundance between smokers and non-smokers in both Task 2 and Task 3.

 In differential abundance analysis conducted using Calour, random permutations of sample labels are used to calculate effect sizes and p-values. These permutations introduce randomness into the process. Thereby, this could be a major reason why results obtained in Task 3 are not the same as the results obtained in Task 2.